reactioncommerce/reaction
# feat: Product/Variant GQL fixes

Resolves #6066   
Impact: **critical**  
Type: **feature|bugfix**

## Issue

There are some missing things for the product and variant GraphQL resolvers that make them difficult to use for administration.

- New variants are usually hidden. Variant resolvers only show visible, to its not possible to change this status unless you know the ID
- Media field is missing from the variants and options
- The media object is missing the ID field which is needed for deletions
- Some translations are missing for UI elements

## Solution

Add the missing things listed below:

- Allow for querying of hidden and archived variants.
- Add media resolver to `variants` and `options` resolvers
- Add missing `_id` field to `ImageInfo` type.
- Add missing translations

## Breaking changes

none

## Testing
Ensure integration tests pass and look good. 

**Note, Media is mocked because there is currently no mock in the test suite for it. Just making sure if media is present, no matter the mongo selector, that it does appear in the response with correct info.**

1. Query a product using the query below
1. Include `shouldIncludeHiddenVariants` and `shouldIncludeArchivedVariants` for some queries.
1. Remove `shouldIncludeHiddenVariants` and `shouldIncludeArchivedVariants`  for some queries. You should see both `hidden` and `visible` products, but not see any `archived/deleted` variants.
1. Repeat the above steps for `shouldIncludeHiddenOptions` and `shouldIncludeArchivedOptions`

Variables.

```js
{
  "productId": "<PRODUCT_ID>",
  "shopId": "<SHOP_ID>",
  "shouldIncludeHiddenVariants": true,
  "shouldIncludeArchivedVariants": false,
  "shouldIncludeHiddenOptions": true,
  "shouldIncludeArchivedOptions": false
}
```

Query
```graphql
query Product(
  $productId: ID!
  $shopId: ID!
  $shouldIncludeHiddenVariants: Boolean
  $shouldIncludeArchivedVariants: Boolean
  $shouldIncludeHiddenOptions: Boolean
  $shouldIncludeArchivedOptions: Boolean
) {
  product(
    productId: $productId
    shopId: $shopId
  ) {
    _id
    title
    media {
      _id
      URLs {
        small
        medium
        large
        thumbnail
      }
    }
    variants(
      shouldIncludeHidden: $shouldIncludeHiddenVariants,
      shouldIncludeArchived: $shouldIncludeArchivedVariants
    ) {
      _id
      title
      media {
        _id
        URLs {
          small
          medium
          large
          thumbnail
        }
      }
      options(
        shouldIncludeHidden: $shouldIncludeHiddenOptions,
        shouldIncludeArchived: $shouldIncludeArchivedOptions
      ) {
        _id
        title
        media {
          _id
          URLs {
            small
            medium
            large
            thumbnail
          }
        }
      }
    }
  }
}
```
